moshi-adapters/src/main/java/com/squareup/moshi/adapters/Iso8601Utils.kt

```diff
@@ -0,0 +1,271 @@
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi.adapters
+
+import com.squareup.moshi.JsonDataException
+import java.util.Calendar
+import java.util.Date
+import java.util.GregorianCalendar
+import java.util.Locale
+import java.util.TimeZone
+import kotlin.math.min
+import kotlin.math.pow
+
+/*
+ * Jackson’s date formatter, pruned to Moshi's needs. Forked from this file:
+ * https://github.com/FasterXML/jackson-databind/blob/67ebf7305f492285a8f9f4de31545f5f16fc7c3a/src/main/java/com/fasterxml/jackson/databind/util/ISO8601Utils.java
+ *
+ * Utilities methods for manipulating dates in iso8601 format. This is much much faster and GC
+ * friendly than using SimpleDateFormat so highly suitable if you (un)serialize lots of date
+ * objects.
+ *
+ * Supported parse format:
+ * `[yyyy-MM-dd|yyyyMMdd][T(hh:mm[:ss[.sss]]|hhmm[ss[.sss]])]?[Z|[+-]hh[:]mm]]`
+ *
+ * @see [this specification](http://www.w3.org/TR/NOTE-datetime)
+ */
+
+/** ID to represent the 'GMT' string  */
+private const val GMT_ID = "GMT"
+
+/** The GMT timezone, prefetched to avoid more lookups.  */
+private val TIMEZONE_Z: TimeZone = TimeZone.getTimeZone(GMT_ID)
+
+/** Returns `date` formatted as yyyy-MM-ddThh:mm:ss.sssZ  */
+internal fun Date.formatIsoDate(): String {
+  val calendar: Calendar = GregorianCalendar(TIMEZONE_Z, Locale.US)
+  calendar.time = this
+
+  // estimate capacity of buffer as close as we can (yeah, that's pedantic ;)
+  val capacity = "yyyy-MM-ddThh:mm:ss.sssZ".length
+  val formatted = StringBuilder(capacity)
+  padInt(formatted, calendar[Calendar.YEAR], "yyyy".length)
+  formatted.append('-')
+  padInt(formatted, calendar[Calendar.MONTH] + 1, "MM".length)
+  formatted.append('-')
+  padInt(formatted, calendar[Calendar.DAY_OF_MONTH], "dd".length)
+  formatted.append('T')
+  padInt(formatted, calendar[Calendar.HOUR_OF_DAY], "hh".length)
+  formatted.append(':')
+  padInt(formatted, calendar[Calendar.MINUTE], "mm".length)
+  formatted.append(':')
+  padInt(formatted, calendar[Calendar.SECOND], "ss".length)
+  formatted.append('.')
+  padInt(formatted, calendar[Calendar.MILLISECOND], "sss".length)
+  formatted.append('Z')
+  return formatted.toString()
+}
+
+/**
+ * Parse a date from ISO-8601 formatted string. It expects a format
+ * `[yyyy-MM-dd|yyyyMMdd][T(hh:mm[:ss[.sss]]|hhmm[ss[.sss]])]?[Z|[+-]hh:mm]]`
+ *
+ * @receiver ISO string to parse in the appropriate format.
+ * @return the parsed date
+ */
+internal fun String.parseIsoDate(): Date {
+  return try {
+    var offset = 0
+
+    // extract year
+    val year = parseInt(this, offset, 4.let { offset += it; offset })
```

Comment:
> `4.let { offset += it; offset }`

oof, that takes a second to grok, but I can't think of a better oneliner 🤷🏾

Reply:
Would `run { offset += 4; offset }` work?

Reply:
I thought about that, it's nice that `4.` is at the front of the `let` construction though

Reply:
yeah it's a little wonky and one of those places that you can't write it the same way in kotlin